File: src/slc_atlas/models/browser.py
# SPDX-FileCopyrightText: 2026 Dong Lab, Yale School of Medicine <https://donglab.org>
#
# SPDX-License-Identifier: Apache-2.0

from typing import Literal

from pydantic import BaseModel

ChromRole = Literal["primary", "alt", "unplaced"]


class Chrom(BaseModel):
    chrom: str
    ensembl: str
    size: int
    role: ChromRole


class CoverageTrack(BaseModel):
    track_id: str
    label: str
    group: str
    stranded: bool
    local: bool
    file: str | None = None
    plus_file: str | None = None
    minus_file: str | None = None
    url: str | None = None
    plus_url: str | None = None
    minus_url: str | None = None
    bin: int = 0
    bytes: int = 0
    # Minimum bases per column for trusted summaries, or 0 when summaries are unavailable
    reduction: int = 0
    chroms: list[str] = []


class GwasStudy(BaseModel):
    study_id: str
    trait: str
    citation: str | None = None
    source: str | None = None
    assembly: str | None = None
    license_spdx: str | None = None
    n_variants: int = 0
    # Chromosomes represented by the study, used to distinguish unavailable data from no hits
    chroms: list[str] = []
    # GWAS pyramid bin sizes in bases per column, finest first
    levels: list[int] = []


class TrackManifest(BaseModel):
    chroms: list[Chrom]
    tracks: list[CoverageTrack]
    studies: list[GwasStudy]


class Region(BaseModel):
    gene_id: str
    symbol: str | None = None
    chrom: str
    chrom_ensembl: str
    chrom_size: int | None = None
    gene_start: int
    gene_end: int
    strand: str | None = None
    window_start: int
    window_end: int
    # How far the view may travel from this gene, which is as much genome as the build kept
    pan_start: int
    pan_end: int
    flank: int
    clipped: bool = False

File: web/src/types/browser.ts
// SPDX-FileCopyrightText: 2026 Dong Lab, Yale School of Medicine <https://donglab.org>
//
// SPDX-License-Identifier: Apache-2.0

// Mirrors src/atlasforge/models/browser.py using 0-based, half-open coordinates

export type ChromRole = "primary" | "alt" | "unplaced"

export interface Chrom {
  chrom: string
  ensembl: string
  size: number
  role: ChromRole
}

export interface CoverageTrack {
  track_id: string
  label: string
  group: string
  stranded: boolean
  local: boolean
  file: string | null
  plus_file: string | null
  minus_file: string | null
  url: string | null
  plus_url: string | null
  minus_url: string | null
  bin: number
  bytes: number
  // Bases per column past which this track's own summaries may be drawn, 0 where they may not be
  reduction: number
  chroms: string[]
}

export interface GwasStudy {
  study_id: string
  trait: string
  citation: string | null
  source: string | null
  assembly: string | null
  license_spdx: string | null
  n_variants: number
  // Distinguishes chromosomes without study data from those without significant hits
  chroms: string[]
  // GWAS pyramid bin sizes in bases per column, finest first
  levels: number[]
}

export interface TrackManifest {
  chroms: Chrom[]
  tracks: CoverageTrack[]
  studies: GwasStudy[]
}

export interface BrowserGene {
  gene_id: string
  symbol: string
  name: string | null
  chrom: string
  is_atlas: boolean
  window_start: number
  window_end: number
}

export interface Exon {
  start: number
  end: number
}

export interface TranscriptModel {
  transcript_id: string
  transcript_version: string | null
  model_gene_id: string | null
  gene_name: string | null
  biotype: string | null
  chrom: string
  start: number
  end: number
  strand: string
  cds_start: number | null
  cds_end: number | null
  exons: Exon[]
  is_atlas_gene: boolean
}

export interface Region {
  gene_id: string
  symbol: string | null
  chrom: string
  chrom_ensembl: string
  chrom_size: number | null
  gene_start: number
  gene_end: number
  strand: string | null
  window_start: number
  window_end: number
  // How far the view may travel from this gene, which is as much genome as the build kept
  pan_start: number
  pan_end: number
  flank: number
  clipped: boolean
}

// One bigBed row: where it is, and one tab-separated string of everything else
export interface RawFeature {
  start: number
  end: number
  rest: string
}
